dwave/system/composites/linear_ancilla.py

```diff
@@ -0,0 +1,214 @@
+    accepted by the child sampler and parameters added by the composite.
+    """
+
+    properties = None  # overwritten by init
+    """dict: Properties in the form of a dict.
+
+    Contains the properties of the child sampler.
+    """
+
+    @nonblocking_sample_method
+    def sample(
+        self,
+        bqm: dimod.BinaryQuadraticModel,
+        *,
+        h_tolerance: numbers.Number = 0,
+        default_flux_bias_range: Sequence[float] = [-0.005, 0.005],
+        **parameters,
+    ):
+        """Sample from the provided binary quadratic model.
+
+        .. note::
+            This composite does not suport the auto_scale parameter. BQM Scaling
+            can be done with :class:`dimod.ScaleComposite`.
+
+        Args:
+            bqm (:obj:`~dimod.BinaryQuadraticModel`):
+                Binary quadratic model to be sampled from.
+
+            h_tolerance (:class:`numbers.Number`):
+                Magnitude of the linear bias can be left on the qubit. Assumed to be positive. Defaults to zero.
+
+            default_flux_bias_range (:class:`typing.Sequence`):
+                Flux bias range safely accepted by the QPU, the larger the better.
+
+            **parameters:
+                Parameters for the sampling method, specified by the child
+                sampler.
+
+        Returns:
+            :obj:`~dimod.SampleSet`
+
+        """
+        if h_tolerance < 0:
+            raise ValueError("h_tolerance needs to be positive or zero")
+
+        child = self.child
+        qpu_properties = innermost_child_properties(child)
+        g_target = child.to_networkx_graph()
+        g_source = dimod.to_networkx_graph(bqm)
+        j_range = qpu_properties["extended_j_range"]
+        flux_bias_range = qpu_properties.get("flux_bias_range", default_flux_bias_range)
+
+        # Positive couplings tend to have smaller control error,
+        # we default to them if they have the same magnitude than negative couplings
+        # https://docs.dwavesys.com/docs/latest/c_qpu_ice.html#overview-of-ice
+        largest_j = j_range[1] if abs(j_range[1]) >= abs(j_range[0]) else j_range[0]
+        largest_j_sign = np.sign(largest_j)
+
+        # To implement the bias sign through flux bias sign,
+        # we pick a range (magnitude) that we can sign-flip
+        fb_magnitude = min(abs(b) for b in flux_bias_range)
+        flux_biases = [0] * qpu_properties["num_qubits"]
+
+        _bqm = bqm.copy()
+        used_ancillas = defaultdict(list)
+        for variable, bias in bqm.iter_linear():
+            if abs(bias) <= h_tolerance:
+                continue
+            if abs(bias) - h_tolerance > abs(largest_j):
+                return NotImplementedError(
+                    "linear biases larger than the strongest coupling are not supported yet"
+                )  # TODO: implement larger biases through multiple ancillas
+
+            available_ancillas = set(g_target.adj[variable]) - set(g_source.nodes())
```

Comment:
I believe we can just do 
```suggestion
            available_ancillas = set(g_target.adj[variable]) - g_source.nodes()
```
which avoids a semi-expensive copy. Elsewhere as well. But I might be misremembering.

Reply:
agreed. same for line 143 `set(used_ancillas)` --> `used_ancillas.keys()`